### PR TITLE
Add usbip module for usb support in WSL

### PR DESCRIPTION
**Pull request by terlar**
This adds a module to enable USB support via usbip together with a [usbip-win](https://github.com/dorssel/usbipd-win) running on the host.

- Adds the usbip package to be able to run adhoc usbip command e.g. (usbip list --remote 172.23.96.1)
- Enables udev
- Creates systemd services to auto-attach USB devices based on busid

One caveat with accessing USB devices is that they will require wider permissions than that will most likely be configured. For example the infrastructure provided by NixOS to enable YubiKeys will add a rule with permission mode `0660`. While when running WSL with usbip you need to set `0666` in the udev rules.

So for example to get YubiKeys working you would need something like this:
```nix
{
  services = {
    pcscd.enable = true;
    udev = {
      packages = [pkgs.yubikey-personalization];
      extraRules = lib.optionalString config.wsl.usbip.enable ''
        SUBSYSTEM=="usb", ATTRS{idVendor}=="1050", ATTRS{idProduct}=="0010|0110|0111|0114|0116|0401|0403|0405|0407|0410", MODE="0666"
      '';
    };
  };
}
```

If we look into the udev rule provided by the `yubikey-personalization` package it looks like this:
```
$ nix build nixpkgs#yubikey-personalization
$ cat result/lib/udev/rules.d/69-yubikey.rules
ACTION!="add|change", GOTO="yubico_end"

# Udev rules for letting the console user access the Yubikey USB
# device node, needed for challenge/response to work correctly.

# Yubico Yubikey II
ATTRS{idVendor}=="1050", ATTRS{idProduct}=="0010|0110|0111|0114|0116|0401|0403|0405|0407|0410", \
    ENV{ID_SECURITY_TOKEN}="1"

LABEL="yubico_end"
```

I don't know enough about udev rules, but would it be possible to do something generic like this:
```
SUBSYSTEM=="usb", MODE="0666"
```

Any ideas appreciated. At least with this configuration I have YubiKey's working within WSL with auto-attach, so I just plug in/unplug and can use it like I would regularly expect on a normal Linux device.

### CODE REVIEW

File: modules/usbip.nix
Comment (SuperSandro2000) [inline modules/usbip.nix:7]: Can we use the release tags?

File: modules/usbip.nix
Comment (SuperSandro2000) [inline modules/usbip.nix:24]:
```suggestion
  config = mkIf cfg.enable {
```
or do we have similar checks on other modules?

File: modules/usbip.nix
Comment (SuperSandro2000) [inline modules/usbip.nix:25]:
```suggestion
```

File: modules/usbip.nix
Comment (SuperSandro2000) [inline modules/usbip.nix:40]: Aren't those there by default?

File: modules/usbip.nix
Comment (terlar) [reply modules/usbip.nix:24]: Yeah, that is where I got it, I'm not sure how likely it is that someone configures the sub option but not the top one. But it could allow for some kind of composition/partial configuration. Thought it had a potential use case. But I'm fine either way

File: modules/usbip.nix
Comment (terlar) [reply modules/usbip.nix:40]: You might be correct. I'll try to remove them and see if the script executes.

**Comment by nzbr** [comment]
Why do devices need to have `0666` permissions? They don't need them on bare-metal NixOS, so why would they need them here? Shouldn't it be possible to add the user to the appropriate groups instead?

**Comment by terlar** [comment]
Okay, this is a bit strange, so I investigated a bit how it works on a pure NixOS machine as opposed to NixOS in WSL.

On WSL my USB-device gets:
```
crw-rw---- 1 root root 189,   2 Jun 26 02:24 /dev/bus/usb/001/003
```

However on NixOS for the same device I get:
```
crw-rw-rw-+ 1 root root 189,   2 Jun 26 02:24 /dev/bus/usb/001/018
```

So somehow the pure NixOS version applied some udev-rule to make that device writeable.

On pure NixOS it also appeared under `/dev/input` which didn't exist at all with WSL.

Another option I checked was that I can assign a group like so:
```
SUBSYSTEM=="usb", GROUP="usb"
```

But then how it works would depart. I will investigate a bit more to see if I can find out why the udev rules evaluate differently.

**Comment by terlar** [comment]
On bare-metal NixOS it has the "same" permissions, but it has a ACL associated with the usb device.
```
$ ls -la /dev/bus/usb/001/007
crw-rw----+ 1 root root 189,   2 Jun 26 02:24 /dev/bus/usb/001/007
$ getfacl /dev/bus/usb/001/007
# file: dev/bus/usb/001/007
# owner: root
# group: root
user::rw-
user:myuser:rw-
group::rw-
mask::rw-
other::r--
```

I'm new to the ACL:s and just found out about them now. But that seems to be the difference.

File: modules/usbip.nix
Comment (terlar) [reply modules/usbip.nix]: A release have not happened, so I can use the release tag.

**Comment by terlar** [comment]
@nzbr and @SuperSandro2000, any ideas how to proceed? The script has now received a tag, so I have updated that. Other than that I guess it is a few open questions. I also noticed a test failed, but I am not sure if it is related to my changes?

**Comment by SuperSandro2000** [comment]
The test failure is related to docker being broken on the commit on stable.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by nzbr** [comment]
@terlar Have you figured something out regarding udev? If there's no other easy way, giving all USB devices `0666` sounds like the best solution

**Comment by terlar** [comment]
@nzbr No unfortunately not, only the findings I mentioned recently. So the permissions are the same on NixOS standalone and NixOS in WSL. But NixOS standalone has ACL configured for the user so the user gets access. Not sure what is configuring the ACL and why they wouldn't be configured inside WSL.

Either we leave it to the users or give the 0666 permissions to all USB devices. Up to you, I'm fine either way. I guess you are not familiar with the ACL stuff either?

**Review by SuperSandro2000 — APPROVED** [review]
.

**Comment by terlar** [comment]
Seems there was another conflict due to recent changes in `modules/default.nix`. I rebased on the main branch again.

I'm okay to go either way, either like this is or with modifications. Just let me know what to do :)

**Comment by LGUG2Z** [comment]
FWIW I've been using the additions on this branch with the suggested `udev` rule without any problems for a week or so.

For people who go a step further and build a custom WSL kernel with HIDRAW enabled to by able to use FIDO2-related features, this is the extra `udev` rule I added to allow me to use the FIDO2-related `ykman` commands (+FIDO2 SSH) without `sudo`:

```
KERNEL=="hidraw*", SUBSYSTEM=="hidraw", TAG+="uaccess", MODE="0666"
```

**Comment by 573** [comment]
What is the state of affairs for this ?

**Comment by terlar** [comment]
I guess it is waiting for decision on proceed or if there is a way to not have those udev rules or perhaps add them by default.

**Comment by 573** [comment]
For me the question was just how to make use of all of this from the user side as depite all of the guides it was still not entirely made clear how to attach usb devices.

Right I'm using @LGUG2Z 's sumup to get a working module on the nixos side with the udev rules @terlar provided in the PR, still the following intel was missing for me which give still not a full-blown readme style explaination:

At least it seems to be a one-time needed thing: `usbipd bind --busid=1-1 -f` (on further calls with different devices I left the `-f` switch off and it still worked) in admin powershell. If I don't do that, the `usbip list` (in WSL VM) call below gets to answer `usbip: info: no exportable devices found on <ip>`. I can prove in any powershell session (even non-admin) that it worked by saying `usbipd list`: Next to the device in question something like `Attached` (to be fair in my case again I said `usbipd wsl list` and it answered `Attached - non-WSL` then, cool so it is possible to bind devices without providing an instance of WSL) should appear.

On NixOS side: `usbip list --remote="$(grep nameserver /etc/resolv.conf | cut -d' ' -f2)"` (see usb-win's `auto-attach` / @terlar 's PR) is the magic incantation that gives all I need to proceed for now.

Still not clear for me is why my usb scanner isn't `scanimage -L` listed but that belongs to another discussion frankly.

**Pull request merged**
